### PR TITLE
StructuredPruningModifier (channel/filter pruning + compression)

### PR DESCRIPTION
**Pull request by bfineran**
* Introduces `StructuredPruningModifier` that applies filter or channel pruning to a module with the option to coordinate masks by parameter groups
* Introduces `LayerThinningModifier` that takes in a static mapping generated from `sparseml.onnx.optim. get_param_structured_pruning_group_dependencies` describing structured pruning dependencies that removes structurally  pruned dimensions and their dependencies from the model
* includes a fix to the torch ONNX export flow to keep consistent batch norm parameter names
* manually tested in a one-shot flow with `apply` and training-aware using a simple recipe and the pytorch integration script
* *known training bug* - since after compression, the graph "resets" to dense, pruning will compound at every pruning step, (ie a 30% prune target in 6 steps may produce a 75% pruned model etc). working on a fix
* unit tests updated, and e2e tests for one shot structured pruning with mobilenet and resnet50 added to the test suite

Example resnet50 recipe: (uncomment `param_groups` and set `strict` to `True` to prune and compress along param groups)

```yaml
version: 1.1.0

# Modifiers Groups:
training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0.0
    end_epoch: 3.0

  - !SetLearningRateModifier
    start_epoch: 0.0
    learning_rate: 0.008

pruning_modifiers:
  - !StructuredPruningModifier
    end_epoch: 2.0
    final_sparsity: 0.3
    init_sparsity: 0.0
    inter_func: cubic
    leave_enabled: True
    log_types: __ALL__
    mask_type: filter
    # param_groups: [['sections.0.2.conv3.weight', 'sections.0.1.conv3.weight', 'sections.0.0.identity.conv.weight', 'sections.0.0.conv3.weight'], ['sections.1.3.conv3.weight', 'sections.1.2.conv3.weight', 'sections.1.1.conv3.weight', 'sections.1.0.identity.conv.weight', 'sections.1.0.conv3.weight'], ['sections.2.5.conv3.weight', 'sections.2.4.conv3.weight', 'sections.2.3.conv3.weight', 'sections.2.2.conv3.weight', 'sections.2.1.conv3.weight', 'sections.2.0.identity.conv.weight', 'sections.2.0.conv3.weight']]
    params: ['input.conv.weight', 'sections.0.0.conv1.weight', 'sections.0.0.conv2.weight', 'sections.0.0.conv3.weight', 'sections.0.0.identity.conv.weight', 'sections.0.1.conv1.weight', 'sections.0.1.conv2.weight', 'sections.0.1.conv3.weight', 'sections.0.2.conv1.weight', 'sections.0.2.conv2.weight', 'sections.0.2.conv3.weight', 'sections.1.0.conv1.weight', 'sections.1.0.conv2.weight', 'sections.1.0.conv3.weight', 'sections.1.0.identity.conv.weight', 'sections.1.1.conv1.weight', 'sections.1.1.conv2.weight', 'sections.1.1.conv3.weight', 'sections.1.2.conv1.weight', 'sections.1.2.conv2.weight', 'sections.1.2.conv3.weight', 'sections.1.3.conv1.weight', 'sections.1.3.conv2.weight', 'sections.1.3.conv3.weight', 'sections.2.0.conv1.weight', 'sections.2.0.conv2.weight', 'sections.2.0.conv3.weight', 'sections.2.0.identity.conv.weight', 'sections.2.1.conv1.weight', 'sections.2.1.conv2.weight', 'sections.2.1.conv3.weight', 'sections.2.2.conv1.weight', 'sections.2.2.conv2.weight', 'sections.2.2.conv3.weight', 'sections.2.3.conv1.weight', 'sections.2.3.conv2.weight', 'sections.2.3.conv3.weight', 'sections.2.4.conv1.weight', 'sections.2.4.conv2.weight', 'sections.2.4.conv3.weight', 'sections.2.5.conv1.weight', 'sections.2.5.conv2.weight', 'sections.2.5.conv3.weight', 'sections.3.0.conv1.weight', 'sections.3.0.conv2.weight', 'sections.3.1.conv1.weight', 'sections.3.1.conv2.weight', 'sections.3.2.conv1.weight', 'sections.3.2.conv2.weight']
    phased: False
    score_type: magnitude
    start_epoch: 0
    update_frequency: 0.5

  - !LayerThinningModifier
    strict: False
    start_epoch: 0.0
    update_epochs: [1.0, 2.0]
    structure_type: filter
    param_group_dependency_map: {'input.conv.weight': ['sections.0.0.identity.conv.weight', 'input.bn.bias', 'input.bn.weight', 'sections.0.0.conv1.weight', 'input.bn.running_mean', 'input.bn.running_var'], 'sections.0.0.conv1.weight': ['sections.0.0.bn1.running_var', 'sections.0.0.bn1.bias', 'sections.0.0.bn1.running_mean', 'sections.0.0.conv2.weight', 'sections.0.0.bn1.weight'], 'sections.0.0.conv2.weight': ['sections.0.0.bn2.bias', 'sections.0.0.bn2.weight', 'sections.0.0.bn2.running_var', 'sections.0.0.conv3.weight', 'sections.0.0.bn2.running_mean'], 'sections.0.1.conv1.weight': ['sections.0.1.conv2.weight', 'sections.0.1.bn1.weight', 'sections.0.1.bn1.running_var', 'sections.0.1.bn1.bias', 'sections.0.1.bn1.running_mean'], 'sections.0.1.conv2.weight': ['sections.0.1.bn2.weight', 'sections.0.1.bn2.running_mean', 'sections.0.1.conv3.weight', 'sections.0.1.bn2.bias', 'sections.0.1.bn2.running_var'], 'sections.0.2.conv1.weight': ['sections.0.2.bn1.running_mean', 'sections.0.2.bn1.running_var', 'sections.0.2.bn1.weight', 'sections.0.2.conv2.weight', 'sections.0.2.bn1.bias'], 'sections.0.2.conv2.weight': ['sections.0.2.bn2.weight', 'sections.0.2.bn2.running_mean', 'sections.0.2.bn2.running_var', 'sections.0.2.bn2.bias', 'sections.0.2.conv3.weight'], 'sections.0.2.conv3.weight,sections.0.1.conv3.weight,sections.0.0.identity.conv.weight,sections.0.0.conv3.weight': ['sections.1.0.conv1.weight', 'sections.0.1.bn3.running_mean', 'sections.0.0.bn3.weight', 'sections.0.0.identity.bn.weight', 'sections.0.0.identity.bn.bias', 'sections.1.0.identity.conv.weight', 'sections.0.1.bn3.running_var', 'sections.0.2.bn3.running_var', 'sections.0.0.bn3.running_mean', 'sections.0.2.conv1.weight', 'sections.0.1.bn3.bias', 'sections.0.2.bn3.running_mean', 'sections.0.0.identity.bn.running_var', 'sections.0.1.conv1.weight', 'sections.0.0.bn3.running_var', 'sections.0.0.identity.bn.running_mean', 'sections.0.0.bn3.bias', 'sections.0.2.bn3.weight', 'sections.0.2.bn3.bias', 'sections.0.1.bn3.weight'], 'sections.1.0.conv1.weight': ['sections.1.0.bn1.running_mean', 'sections.1.0.bn1.running_var', 'sections.1.0.conv2.weight', 'sections.1.0.bn1.bias', 'sections.1.0.bn1.weight'], 'sections.1.0.conv2.weight': ['sections.1.0.bn2.bias', 'sections.1.0.bn2.weight', 'sections.1.0.bn2.running_mean', 'sections.1.0.bn2.running_var', 'sections.1.0.conv3.weight'], 'sections.1.1.conv1.weight': ['sections.1.1.bn1.bias', 'sections.1.1.conv2.weight', 'sections.1.1.bn1.running_var', 'sections.1.1.bn1.weight', 'sections.1.1.bn1.running_mean'], 'sections.1.1.conv2.weight': ['sections.1.1.bn2.running_mean', 'sections.1.1.conv3.weight', 'sections.1.1.bn2.weight', 'sections.1.1.bn2.running_var', 'sections.1.1.bn2.bias'], 'sections.1.2.conv1.weight': ['sections.1.2.bn1.running_mean', 'sections.1.2.bn1.bias', 'sections.1.2.bn1.running_var', 'sections.1.2.conv2.weight', 'sections.1.2.bn1.weight'], 'sections.1.2.conv2.weight': ['sections.1.2.bn2.bias', 'sections.1.2.conv3.weight', 'sections.1.2.bn2.running_mean', 'sections.1.2.bn2.weight', 'sections.1.2.bn2.running_var'], 'sections.1.3.conv1.weight': ['sections.1.3.bn1.running_mean', 'sections.1.3.bn1.weight', 'sections.1.3.bn1.running_var', 'sections.1.3.conv2.weight', 'sections.1.3.bn1.bias'], 'sections.1.3.conv2.weight': ['sections.1.3.bn2.weight', 'sections.1.3.bn2.bias', 'sections.1.3.bn2.running_mean', 'sections.1.3.conv3.weight', 'sections.1.3.bn2.running_var'], 'sections.1.3.conv3.weight,sections.1.2.conv3.weight,sections.1.1.conv3.weight,sections.1.0.identity.conv.weight,sections.1.0.conv3.weight': ['sections.1.2.bn3.running_mean', 'sections.1.0.identity.bn.weight', 'sections.1.1.conv1.weight', 'sections.1.0.bn3.weight', 'sections.1.3.bn3.bias', 'sections.1.1.bn3.weight', 'sections.1.3.bn3.running_var', 'sections.1.3.conv1.weight', 'sections.1.3.bn3.running_mean', 'sections.1.2.bn3.bias', 'sections.1.2.conv1.weight', 'sections.1.1.bn3.running_var', 'sections.2.0.conv1.weight', 'sections.1.0.bn3.running_var', 'sections.1.1.bn3.bias', 'sections.1.3.bn3.weight', 'sections.1.0.identity.bn.running_var', 'sections.1.0.identity.bn.running_mean', 'sections.1.0.bn3.bias', 'sections.1.0.bn3.running_mean', 'sections.2.0.identity.conv.weight', 'sections.1.2.bn3.weight', 'sections.1.1.bn3.running_mean', 'sections.1.2.bn3.running_var', 'sections.1.0.identity.bn.bias'], 'sections.2.0.conv1.weight': ['sections.2.0.bn1.bias', 'sections.2.0.conv2.weight', 'sections.2.0.bn1.running_var', 'sections.2.0.bn1.weight', 'sections.2.0.bn1.running_mean'], 'sections.2.0.conv2.weight': ['sections.2.0.bn2.bias', 'sections.2.0.bn2.running_var', 'sections.2.0.conv3.weight', 'sections.2.0.bn2.running_mean', 'sections.2.0.bn2.weight'], 'sections.2.1.conv1.weight': ['sections.2.1.bn1.running_var', 'sections.2.1.bn1.running_mean', 'sections.2.1.conv2.weight', 'sections.2.1.bn1.weight', 'sections.2.1.bn1.bias'], 'sections.2.1.conv2.weight': ['sections.2.1.conv3.weight', 'sections.2.1.bn2.running_var', 'sections.2.1.bn2.bias', 'sections.2.1.bn2.weight', 'sections.2.1.bn2.running_mean'], 'sections.2.2.conv1.weight': ['sections.2.2.bn1.bias', 'sections.2.2.bn1.running_mean', 'sections.2.2.conv2.weight', 'sections.2.2.bn1.weight', 'sections.2.2.bn1.running_var'], 'sections.2.2.conv2.weight': ['sections.2.2.conv3.weight', 'sections.2.2.bn2.running_mean', 'sections.2.2.bn2.running_var', 'sections.2.2.bn2.bias', 'sections.2.2.bn2.weight'], 'sections.2.3.conv1.weight': ['sections.2.3.conv2.weight', 'sections.2.3.bn1.bias', 'sections.2.3.bn1.running_mean', 'sections.2.3.bn1.weight', 'sections.2.3.bn1.running_var'], 'sections.2.3.conv2.weight': ['sections.2.3.bn2.running_mean', 'sections.2.3.bn2.bias', 'sections.2.3.bn2.running_var', 'sections.2.3.bn2.weight', 'sections.2.3.conv3.weight'], 'sections.2.4.conv1.weight': ['sections.2.4.bn1.running_var', 'sections.2.4.bn1.bias', 'sections.2.4.conv2.weight', 'sections.2.4.bn1.weight', 'sections.2.4.bn1.running_mean'], 'sections.2.4.conv2.weight': ['sections.2.4.bn2.weight', 'sections.2.4.conv3.weight', 'sections.2.4.bn2.running_mean', 'sections.2.4.bn2.running_var', 'sections.2.4.bn2.bias'], 'sections.2.5.conv1.weight': ['sections.2.5.bn1.weight', 'sections.2.5.bn1.running_mean', 'sections.2.5.bn1.running_var', 'sections.2.5.conv2.weight', 'sections.2.5.bn1.bias'], 'sections.2.5.conv2.weight': ['sections.2.5.bn2.weight', 'sections.2.5.bn2.bias', 'sections.2.5.conv3.weight', 'sections.2.5.bn2.running_mean', 'sections.2.5.bn2.running_var'], 'sections.2.5.conv3.weight,sections.2.4.conv3.weight,sections.2.3.conv3.weight,sections.2.2.conv3.weight,sections.2.1.conv3.weight,sections.2.0.identity.conv.weight,sections.2.0.conv3.weight': ['sections.2.4.bn3.running_var', 'sections.2.0.bn3.running_mean', 'sections.2.5.bn3.bias', 'sections.2.3.bn3.bias', 'sections.2.2.conv1.weight', 'sections.2.0.bn3.running_var', 'sections.2.2.bn3.running_var', 'sections.2.0.identity.bn.running_mean', 'sections.2.0.identity.bn.bias', 'sections.2.1.bn3.weight', 'sections.2.0.bn3.bias', 'sections.3.0.identity.conv.weight', 'sections.2.1.bn3.running_mean', 'sections.2.2.bn3.running_mean', 'sections.2.3.bn3.running_mean', 'sections.2.0.identity.bn.weight', 'sections.2.5.bn3.running_var', 'sections.2.1.bn3.bias', 'sections.2.4.bn3.bias', 'sections.2.4.bn3.weight', 'sections.2.0.bn3.weight', 'sections.2.2.bn3.bias', 'sections.2.2.bn3.weight', 'sections.2.1.bn3.running_var', 'sections.2.3.conv1.weight', 'sections.2.3.bn3.weight', 'sections.2.5.conv1.weight', 'sections.2.5.bn3.weight', 'sections.2.5.bn3.running_mean', 'sections.2.1.conv1.weight', 'sections.2.4.bn3.running_mean', 'sections.2.3.bn3.running_var', 'sections.3.0.conv1.weight', 'sections.2.4.conv1.weight', 'sections.2.0.identity.bn.running_var'], 'sections.3.0.conv1.weight': ['sections.3.0.bn1.bias', 'sections.3.0.bn1.running_mean', 'sections.3.0.conv2.weight', 'sections.3.0.bn1.running_var', 'sections.3.0.bn1.weight'], 'sections.3.0.conv2.weight': ['sections.3.0.bn2.running_var', 'sections.3.0.bn2.bias', 'sections.3.0.conv3.weight', 'sections.3.0.bn2.weight', 'sections.3.0.bn2.running_mean'], 'sections.3.1.conv1.weight': ['sections.3.1.bn1.running_var', 'sections.3.1.bn1.bias', 'sections.3.1.bn1.weight', 'sections.3.1.conv2.weight', 'sections.3.1.bn1.running_mean'], 'sections.3.1.conv2.weight': ['sections.3.1.bn2.running_var', 'sections.3.1.bn2.weight', 'sections.3.1.conv3.weight', 'sections.3.1.bn2.bias', 'sections.3.1.bn2.running_mean'], 'sections.3.2.conv1.weight': ['sections.3.2.bn1.running_var', 'sections.3.2.bn1.weight', 'sections.3.2.bn1.bias', 'sections.3.2.bn1.running_mean', 'sections.3.2.conv2.weight'], 'sections.3.2.conv2.weight': ['sections.3.2.bn2.weight', 'sections.3.2.conv3.weight', 'sections.3.2.bn2.running_mean', 'sections.3.2.bn2.bias', 'sections.3.2.bn2.running_var']}
```

### CODE REVIEW

**Comment by markurtz** [comment]
@bfineran just putting a note in here over what we talked over to get our base case working for uniform structured and the layer thinner

**Comment by markurtz** [comment]
Talked with @bfineran and reconfirmed with @govindr-nm that we want to have a remainder on each input channel to match how the deepsparse engine will run which will pad those input channels to the next 4

**Review by markurtz — APPROVED** [review]
Overall looks good, will need to play around with it to get a sense for any additional features/changes for the future

**Review by kazukageorge — APPROVED** [review]
